sqlsmith: support user-defined composite types in sqlsmith

michae2 opened the pull request:
The included commits add basic support for user-defined composite types to sqlsmith. This is enough to produce tables with columns of composite type that are used in inserts, updates, deletes, and selects. Notably missing are:
- accesses of individual subfields in a composite value (i.e. `(col).field` syntax)
- modification of individual subfields in a composite value, due to https://github.com/cockroachdb/cockroach/issues/102984
- alters of composite types, due to https://github.com/cockroachdb/cockroach/issues/48701

Epic: None

Release note: None

michae2 committed:
sqlsmith: add a composite type column to the `seed` table

Release note: None

michae2 committed:
sqlsmith: support user-defined composite types in `extractTypes`

As part of `ReloadSchemas` sqlsmith gathers all user-defined types in
order to use them in statements. Add support for user-defined composite
types to this.

Also rename `getRandUserDefinedTypeLabel` to `getRandEnumVal`.

Epic: None

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/103081)
<!-- Reviewable:end -->

michae2 commented:
I need to fix some tests, so hold off for now (unless you want a sneak preview).

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> thank you!

> Reviewed 2 of 2 files at r1, 8 of 8 files at r5, 4 of 4 files at r6, 1 of 1 files at r7, 1 of 1 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/103081#-:-O-m25us1HTnm4XvgQ78:bro3093)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @michae2)



<!-- Sent from Reviewable.io -->

yuzefovich reviewed (approved):
Nice! <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

nit: PR labels need an adjustment. I'd lean towards not backporting anywhere initially until we stabilize it in case there is some fallout.

> Reviewed 8 of 8 files at r5, 4 of 4 files at r6, 1 of 1 files at r7, 1 of 1 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/103081#-:-O-xTLk035TGaAZW4zZr:bnc70vk)* status: <span>:shipit:</span> complete! 2 of 0 LGTMs obtained (waiting on @michae2)

___
*[`pkg/internal/sqlsmith/alter.go` line 582 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/103081#-O-xQo6w9hsH_ETx5IaM:-O-xQo6w9hsH_ETx5IaN:bxt6czs) ([raw file](https://github.com/cockroachdb/cockroach/blob/6dec5695bb770ceaefeeccd8b271be56f324be65/pkg/internal/sqlsmith/alter.go#L582)):*
> ```Go
> 
> func makeAlterTypeDropValue(s *Smither) (tree.Statement, bool) {
> 	enumVal, udtName, ok := s.getRandUserDefinedEnumLabel()
> ```

nit: should we do `s/udtName/enumName/` for all callsites of this function?
___
*[`pkg/internal/sqlsmith/schema.go` line 249 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/103081#-O-xQdB5A-ZzxSiFu_cC:-O-xQdB5A-ZzxSiFu_cD:b-17sjxi) ([raw file](https://github.com/cockroachdb/cockroach/blob/6dec5695bb770ceaefeeccd8b271be56f324be65/pkg/internal/sqlsmith/schema.go#L249)):*
> ```Go
> }
> 
> func (s *Smither) getRandUserDefinedEnumLabel() (*tree.EnumValue, *tree.TypeName, bool) {
> ```

nit: should we shorten it to `getRandEnumVal`?


<!-- Sent from Reviewable.io -->

michae2 committed:
randgen: add type resolver argument to PopulateTableWithRandData

Add type resolution to PopulateTableWithRandData. Conveniently, thanks
to the earlier commit adding composite types to `extractTypes`, a
smither can act as a type resolver, so use this in some places. If we
don't have a type resolver simply skip over user-defined types.

Epic: None

Release note: None

michae2 reviewed (commented):
Thanks for the reviews!

This is still waiting on:
- a fix for https://github.com/cockroachdb/cockroach/issues/125550
- a fix for https://github.com/cockroachdb/cockroach/issues/125934

TODO:
- add UDTs to RandCreateTable
- fix a few more test failures

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/103081#-:-O0Ba7HVEA8yj4xcQrfK:bldo9ap)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @msbutler, @rytaft, @wenyihu6, and @yuzefovich)

___
*[`pkg/internal/sqlsmith/alter.go` line 582 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/103081#-O-xQo6w9hsH_ETx5IaM:-O0Ba35UCCaBncwrzULd:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6dec5695bb770ceaefeeccd8b271be56f324be65/pkg/internal/sqlsmith/alter.go#L582)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: should we do `s/udtName/enumName/` for all callsites of this function?
</blockquote></details>

Done.
___
*[`pkg/internal/sqlsmith/schema.go` line 249 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/103081#-O-xQdB5A-ZzxSiFu_cC:-O0Ba4Dd-PRVSLHkIDNQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6dec5695bb770ceaefeeccd8b271be56f324be65/pkg/internal/sqlsmith/schema.go#L249)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: should we shorten it to `getRandEnumVal`?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

rytaft commented:
@michae2 Can we merge this now?

michae2 commented:
> @michae2 Can we merge this now?

Yes, I'll rebase and merge.

michae2 committed:
sqlsmith: use UDTs in random CREATE TABLE statements

Add user-defined types to random CREATE TABLE statements produced by
SQLSmith.

Epic: None

Release note: None

michae2 commented:
I've rebased this on top of #127185 and added another commit that creates composite types in randomly generated tables. Still have to fix a couple more tests.

michae2 commented:
Found a new issue in CI: #138874
